ts/a11y/SpeechMenu.ts

```diff
@@ -0,0 +1,211 @@
+}
+
+/**
+ * Generates the menu items for the base preference menu.
+ *
+ * @param {string} previous The currently set preferences.
+ * @return The menu items.
+ */
+function basePreferences(previous: string) {
+  const items = [
+    {
+      type: 'radio',
+      content: 'No Preferences',
+      id: 'clearspeak-default',
+      variable: 'speechRules'
+    },
+    {
+      type: 'radio',
+      content: 'Current Preferences',
+      id: 'clearspeak-' + previous,
+      variable: 'speechRules'
+    },
+    {
+      type: 'rule'
+    },
+  ];
+  return items;
+}
+
+/**
+ * Generates the items for smart preference choices, depending on the top most
+ *
+ * @param {string} previous The currently set preferences.
+ * @param {string} smart The semantic type of the smart preferences.
+ * @param {string} locale The current locale.
+ * @return The menu of the smart choices.
+ */
+function smartPreferences(previous: string, smart: string, locale: string) {
+  const prefs = Sre.clearspeakPreferences.getLocalePreferences();
+  const loc = prefs[locale];
+  if (!loc) {
+    return [];
+  }
+  const items = [
+    {type: 'label', content: 'Preferences for ' + smart},
+    {type: 'rule'}
+  ];
+  return items.concat(loc[smart].map(function (x) {
+    const [key, value] = x.split('_');
+    return {
+      type: 'radioCompare',
+      content: value,
+      id:
+      'clearspeak-' + Sre.clearspeakPreferences.addPreference(previous, key, value),
+      variable: 'speechRules',
+      comparator: (x: string, y: string) => {
+        if (x === y) {
+          return true;
+        }
+        if (value !== 'Auto') {
+          return false;
+        }
+        let [dom1, pref] = x.split('-');
+        let [dom2] = y.split('-');
+        return dom1 === dom2 && !Sre.clearspeakPreferences.fromPreference(pref)[key];
+      }
+    };
+  }));
+}
+
+/**
+ * Creates dynamic clearspeak menu.
+ * @param {MJContextMenu} menu The context menu.
+ * @param {Submenu} sub The submenu to attach elements to.
+ */
+export function clearspeakMenu(menu: MJContextMenu, sub: Submenu) {
+  let locale = menu.pool.lookup('locale').getValue() as string;
+  const box = csSelectionBox(menu, locale);
+  let items: Object[] = [];
+  let explorer = (menu.mathItem as ExplorerMathItem)?.
+    explorers?.explorers?.speech as SpeechExplorer;
+  if (explorer?.walker) {
+    let semantic = explorer.walker.getFocus()?.getSemanticPrimary();
+    if (semantic) {
+      const previous = Sre.clearspeakPreferences.currentPreference();
+      const smart = Sre.clearspeakPreferences.relevantPreferences(semantic);
+      items = items.concat(basePreferences(previous));
+      items = items.concat(smartPreferences(previous, smart, locale));
+    }
+  }
+  if (box) {
+    items.splice(2, 0, box);
+  }
+  return menu.factory.get('subMenu')(menu.factory, {
+    items: items,
+    id: 'Clearspeak'
+  }, sub);
+}
+MJContextMenu.DynamicSubmenus.set('Clearspeak', clearspeakMenu);
```

Comment:
Same question about dynamic menus here as for the locale menu below.  Do these preferences have to be regenerated for each expression, or can they be reused once created the first time?

Reply:
Does not work here, because the Clearspeak menu is adapted according to the current position of the browser to present the "smart preferences".